src/database_impl/setup.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    borrow::{Borrow, Cow},
+    fmt::Display,
+    sync::{Arc, Mutex},
+};
+
+use async_trait::async_trait;
+use mysql::{prelude::Queryable, Conn, OptsBuilder, Row};
+
+use crate::{config::DatabaseConnConfig, error::Result, Database, SqlnessError};
+
+#[async_trait]
+pub trait DatabaseBuilder: Send + Sync + Default {
+    async fn build(&self, config: DatabaseConnConfig) -> Result<Box<dyn Database>>;
```

Comment:
How about making `Database` (the return type) a type parameter, like
```rust
trait DatabaseBuilder {
    type Database: Database;
    
    async fn build(&self, config: DatabaseConnConfig) -> Result<Self::Database>;
}
```

As the type to be built is determined. And as well as the config type, maybe?